File: id/id-oauth-facebook/src/main/java/io/piano/android/id/facebook/FacebookSignInActivity.java
package io.piano.android.id.facebook;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.Nullable;

import com.facebook.CallbackManager;
import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.login.LoginManager;
import com.facebook.login.LoginResult;

import java.util.Collections;

import io.piano.android.id.PianoOAuthActivity;

public class FacebookSignInActivity extends PianoOAuthActivity {

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        try {
            LoginManager.getInstance().logOut();
            LoginManager.getInstance().logIn(this, Collections.singletonList("email"));
        } catch (Exception ex) {
            setFailureResult(ex);
            finish();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        switch (resultCode) {
            case Activity.RESULT_OK:
                try {
                    final CallbackManager callbackManager = CallbackManager.Factory.create();
                    LoginManager.getInstance().registerCallback(callbackManager, new FacebookCallback<LoginResult>() {
                        @Override
                        public void onSuccess(LoginResult loginResult) {
                            if (!loginResult.getRecentlyGrantedPermissions().contains("email")) {
                                onCancel();
                                return;
                            }
                            LoginManager.getInstance().unregisterCallback(callbackManager);

                            setSuccessResult(FacebookOAuthProvider.NAME, loginResult.getAccessToken().getToken());
                        }

                        @Override
                        public void onCancel() {
                            LoginManager.getInstance().unregisterCallback(callbackManager);
                            setResult(Activity.RESULT_CANCELED);
                        }

                        @Override
                        public void onError(FacebookException e) {
                            LoginManager.getInstance().unregisterCallback(callbackManager);
                            setFailureResult(e);
                        }
                    });

                    callbackManager.onActivityResult(requestCode, resultCode, data);
                } catch (Exception ex) {
                    setFailureResult(ex);
                }
                break;
            case Activity.RESULT_CANCELED:
                setResult(Activity.RESULT_CANCELED);
                break;
        }

        finish();
    }
}

File: id/id-oauth-facebook/src/main/java/io/piano/android/id/facebook/FacebookSignInActivity.kt
package io.piano.android.id.facebook

import android.os.Bundle
import com.facebook.CallbackManager
import com.facebook.FacebookCallback
import com.facebook.FacebookException
import com.facebook.login.LoginManager
import com.facebook.login.LoginResult
import io.piano.android.id.PianoOAuthActivity

class FacebookSignInActivity : PianoOAuthActivity() {
    private val fbCallbackManager = CallbackManager.Factory.create()
    private val callback = object : FacebookCallback<LoginResult> {
        override fun onCancel() {
            setResult(RESULT_CANCELED)
            unregisterCallbacksAndFinish()
        }

        override fun onError(error: FacebookException) {
            setFailureResult(error)
            unregisterCallbacksAndFinish()
        }

        override fun onSuccess(result: LoginResult) {
            if ("email" in result.recentlyGrantedPermissions) {
                setSuccessResult(FacebookOAuthProvider.NAME, result.accessToken.token)
            } else {
                setResult(RESULT_CANCELED)
            }
            unregisterCallbacksAndFinish()
        }
    }

    @Suppress("NOTHING_TO_INLINE")
    private inline fun unregisterCallbacksAndFinish() {
        LoginManager.getInstance().unregisterCallback(fbCallbackManager)
        finish()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        runCatching {
            with(LoginManager.getInstance()) {
                registerCallback(fbCallbackManager, callback)
                logOut()
                logInWithReadPermissions(this@FacebookSignInActivity, fbCallbackManager, listOf("email"))
            }
        }.onFailure {
            setFailureResult(it)
            finish()
        }
    }
}
